File: AllegroDotNet.Wrapper/src/AlTimer.cpp
#include "PCH.h"

#include "Al.h"
#include "AllegroEventSource.h"
#include "AllegroTimer.h"

namespace AllegroDotNet::Wrapper
{
	AllegroEventSource^ Al::GetTimerEventSource(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		auto* nativeTimerEventSource = al_get_timer_event_source(nativeTimer);
		return gcnew AllegroEventSource(nativeTimerEventSource);
	}

	AllegroTimer^ Al::CreateTimer(Double speedSecs)
	{
		auto* nativeTimer = al_create_timer(speedSecs);
		return gcnew AllegroTimer(nativeTimer);
	}

	Boolean Al::GetTimerStarted(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		return al_get_timer_started(nativeTimer);
	}

	Double Al::BpmToSecs(Double bpm)
	{
		return ALLEGRO_BPM_TO_SECS(bpm);
	}

	Double Al::BpsToSecs(Double bps)
	{
		return ALLEGRO_BPS_TO_SECS(bps);
	}

	Double Al::GetTimerSpeed(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		return al_get_timer_speed(nativeTimer);
	}

	Double Al::MSecsToSecs(Double mSecs)
	{
		return ALLEGRO_MSECS_TO_SECS(mSecs);
	}

	Double Al::USecsToSecs(Double uSecs)
	{
		return ALLEGRO_USECS_TO_SECS(uSecs);
	}

	Int64 Al::GetTimerCount(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		return al_get_timer_count(nativeTimer);
	}

	void Al::AddTimerCount(AllegroTimer^ timer, Int64 diff)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_add_timer_count(nativeTimer, diff);
	}

	void Al::DestroyTimer(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_destroy_timer(nativeTimer);
	}

	void Al::ResumeTimer(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_resume_timer(nativeTimer);
	}

	void Al::SetTimerCount(AllegroTimer^ timer, Int64 newCount)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_set_timer_count(nativeTimer, newCount);
	}

	void Al::SetTimerSpeed(AllegroTimer^ timer, Double newSpeedSecs)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_set_timer_speed(nativeTimer, newSpeedSecs);
	}

	void Al::StartTimer(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_start_timer(nativeTimer);
	}

	void Al::StopTimer(AllegroTimer^ timer)
	{
		auto* nativeTimer = timer->GetNativePointer();
		al_stop_timer(nativeTimer);
	}
}

File: Source/AllegroDotNet/Al.Timer.cs
using SubC.AllegroDotNet.Models;
using SubC.AllegroDotNet.Native;

namespace SubC.AllegroDotNet;

/// <summary>
/// This static class contains the Allegro 5 library methods.
/// </summary>
public static partial class Al
{
    public static double USecsToSecs(double x)
    {
        return (x / 1000000.0);
    }

    public static double MSecsToSecs(double x)
    {
        return (x / 1000.0);
    }

    public static double BpsToSecs(double x)
    {
        return (1.0 / x);
    }

    public static double BpmToSecs(double x)
    {
        return (60.0 / x);
    }

    public static AllegroTimer? CreateTimer(double speedSeconds)
    {
        var pointer = Interop.Core.AlCreateTimer(speedSeconds);
        return NativePointer.Create<AllegroTimer>(pointer);
    }

    public static void StartTimer(AllegroTimer? timer)
    {
        Interop.Core.AlStartTimer(NativePointer.Get(timer));
    }

    public static void ResumeTimer(AllegroTimer? timer)
    {
        Interop.Core.AlResumeTimer(NativePointer.Get(timer));
    }

    public static void StopTimer(AllegroTimer? timer)
    {
        Interop.Core.AlStopTimer(NativePointer.Get(timer));
    }

    public static bool GetTimerStarted(AllegroTimer? timer)
    {
        return Interop.Core.AlGetTimerStarted(NativePointer.Get(timer)) != 0;
    }

    public static void DestroyTimer(AllegroTimer? timer)
    {
        Interop.Core.AlDestroyTimer(NativePointer.Get(timer));
    }

    public static long GetTimerCount(AllegroTimer? timer)
    {
        return Interop.Core.AlGetTimerCount(NativePointer.Get(timer));
    }

    public static void SetTimerCount(AllegroTimer? timer, long newCount)
    {
        Interop.Core.AlSetTimerCount(NativePointer.Get(timer), newCount);
    }

    public static void AddTimerCount(AllegroTimer? timer, long diffCount)
    {
        Interop.Core.AlAddTimerCount(NativePointer.Get(timer), diffCount);
    }

    public static double GetTimerSpeed(AllegroTimer? timer)
    {
        return Interop.Core.AlGetTimerSpeed(NativePointer.Get(timer));
    }

    public static void SetTimerSpeed(AllegroTimer? timer, double newSpeedSeconds)
    {
        Interop.Core.AlSetTimerSpeed(NativePointer.Get(timer), newSpeedSeconds);
    }

    public static AllegroEventSource? GetTimerEventSource(AllegroTimer? timer)
    {
        var pointer = Interop.Core.AlGetTimerEventSource(NativePointer.Get(timer));
        return NativePointer.Create<AllegroEventSource>(pointer);
    }
}
